server/src/internalClusterTest/java/org/elasticsearch/search/retriever/RetrieverRewriteIT.java

```diff
@@ -0,0 +1,266 @@
+            }
+            return this;
+        }
+
+        @Override
+        public void extractToSearchSourceBuilder(SearchSourceBuilder sourceBuilder, boolean compoundUsed) {
+            assertNull(sourceBuilder.retriever());
+            innerRetriever.extractToSearchSourceBuilder(sourceBuilder, compoundUsed);
+        }
+
+        @Override
+        public String getName() {
+            return "asserting";
+        }
+
+        @Override
+        protected void doToXContent(XContentBuilder builder, Params params) throws IOException {}
+
+        @Override
+        protected boolean doEquals(Object o) {
+            return false;
+        }
+
+        @Override
+        protected int doHashCode() {
+            return innerRetriever.doHashCode();
+        }
+    }
+
+    private static class AssertingCompoundRetrieverBuilder extends RetrieverBuilder {
+        static final String NAME = "asserting_compound";
+
+        private final String id;
+        private final SetOnce<RetrieverBuilder> innerRetriever;
+
+        public static AssertingCompoundRetrieverBuilder fromXContent(XContentParser parser, RetrieverParserContext context)
+            throws IOException {
+            return PARSER_COMPOUND.apply(parser, context);
+        }
+
+        private AssertingCompoundRetrieverBuilder(String id) {
+            this.id = id;
+            this.innerRetriever = new SetOnce<>(null);
+        }
+
+        private AssertingCompoundRetrieverBuilder(String id, SetOnce<RetrieverBuilder> innerRetriever) {
+            this.id = id;
+            this.innerRetriever = innerRetriever;
+        }
+
+        @Override
+        public boolean isCompound() {
+            return true;
+        }
+
+        @Override
+        public RetrieverBuilder rewrite(QueryRewriteContext ctx) throws IOException {
+            assertNotNull(ctx.getPointInTimeBuilder());
+            assertNull(ctx.convertToInnerHitsRewriteContext());
+            assertNull(ctx.convertToCoordinatorRewriteContext());
+            assertNull(ctx.convertToIndexMetadataContext());
+            assertNull(ctx.convertToSearchExecutionContext());
+            assertNull(ctx.convertToDataRewriteContext());
+            if (innerRetriever.get() != null) {
+                return this;
+            }
+            SetOnce<RetrieverBuilder> innerRetriever = new SetOnce<>();
+            ctx.registerAsyncAction((client, actionListener) -> {
+                SearchSourceBuilder source = new SearchSourceBuilder().pointInTimeBuilder(ctx.getPointInTimeBuilder())
+                    .query(QueryBuilders.termQuery(ID_FIELD, id))
+                    .fetchField(QUERY_FIELD);
+                client.search(new SearchRequest().source(source), new ActionListener<>() {
+                    @Override
+                    public void onResponse(SearchResponse response) {
+                        String query = response.getHits().getAt(0).field(QUERY_FIELD).getValue();
```

Comment:
We should initially enter from the transport search action, so it should be on a transport_worker. Since it's an async action, the thread handling the response can be from another pool. In this case, as it's a search, we might end up responding on the search thread pool of that node. This means the next round of rewrite will also execute on the same thread. This isn't an issue as long as the synchronous work in the rewrite remains light and doesn't perform any IO.